More configurable Aurora

chris-pardy opened the pull request:
We'd like to keep using the builtin Aurora scheduler but we'd like to.
1) Move the default file to a different location, and not have it be named heron.aurora (we're using json for other aurora jobs)
2) Sometimes specify a different file for different topologies.

This PR is intended to enable this workflow by allowing configs to override the location and/or name of the default aurora template. Additionally it looks for a System Property which if set will take precedence. Because this method falls back to the old behavior if the properties aren't set this should be a non-breaking change however there is a very small possibility that someone is running the Heron CLI with the environment property HERON_SCHEDULER_JOB_TEMPLATE set to some garbage. I think the risk is minimal enough that it's not required to require a config be turned on to check the environment property.

kramasamy commented on heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraContext.java:
```diff
@@ -14,16 +14,40 @@
   public static String getJobLinkTemplate(Config config) {
     return config.getStringValue(JOB_LINK_TEMPLATE);
   }
+
```
Instead of using environment variable, a better approach would be to check if these new configs are present, if these are present, use those - otherwise, default to what it is used to be. Alternatively, we can make this config mandatory - with the default pointing to heron.aurora file in the cluster directory.

billonahill commented on heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraContext.java:
```diff
@@ -14,16 +14,40 @@
+   * is used.
+   */
+  public static String getHeronAuroraPath(Config config) {
+    String template = System.getenv(JOB_TEMPLATE_ENV);
```
Instead of mixing both Heron configs and environment variables can we just use configs? Splitting the logic between configs and env variables makes for an inconsistent experience with hidden features. If we use a config param you can still pass it at the command line to override the default:
```
$ heron help submit | grep config-property
  --config-property (key=value; a config key and its value; default: [])
```

chris-pardy replied:
@billonahill I didn't know the --config-property was a thing, thanks. I'll drop the environment config, I actually much prefer the commandline option.

kramasamy replied:
+1 please use --config-property, that is a better approach.

billonahill commented on heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraContext.java:
```diff
@@ -14,16 +14,40 @@
+  public static String getHeronAuroraPath(Config config) {
+    String template = System.getenv(JOB_TEMPLATE_ENV);
+    if (template != null){
+      return new File(template).getPath();
```
We should always join the configured path with the file name for consistency.

kramasamy commented on heron/config/src/yaml/conf/aurora/scheduler.yaml:
```diff
@@ -8,4 +8,10 @@ heron.class.launcher:                        com.twitter.heron.scheduler.aurora.
+#heron.scheduler.job.template:                heron.aurora
+
+# Override the default directory for the Aurora template
+#heron.scheduler.job.template.dir:            /etc/heron
```
By default, this is not installed in /etc/heron. It is installed in where the cluster config resides. Location of where the config resides is provided as configPath in SubmitterMain - you need to use this configPath. Probably, a better approach would be if the config is not present, fall back to the default approach?

chris-pardy replied:
I wanted to put something in this file to indicate that it could be set to a different, fixed directory. The Aurora Context file already does the fall back if the tag is not defined. I thought that having it here but commented out made sense since it would easy to simply uncomment the line and change the directory to make use of it.

kramasamy replied:
sounds good. perhaps, we can make the comment little more verbose 

- in the sense that these two parameters are optional
- Any situations it is used as well

chris-pardy commented:
@kramasamy @billonahill I believe I've addressed all the PR comments. The CI check seems to be failing because I'm missing JavaDoc on a Unit Test that I modified. I'm guessing this is the result of a mis-configured checkstyle rule. Would you prefer if I edited the rule to ignore tests or added JavaDoc to the test?

kramasamy commented on heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraContext.java:
```diff
@@ -14,16 +14,26 @@
+    return new File(config.getStringValue(JOB_TEMPLATE_DIR, Context.heronConf(config)),
+        config.getStringValue(JOB_TEMPLATE,"heron.aurora")).getPath();
+  }
+
```
@chris-pardy - in the case of both the config being absent, how does this work? Is the JOB_TEMPLATE_DIR set to default somewhere?

chris-pardy replied:
The second parameter to the config.getStringValue method is the default value, I'm using the same values here that used to be in AuroraScheduler.

kramasamy commented:
@chris-pardy - meta comment. Do we need two configs for the path? A single config will be sufficient, isn't?

chris-pardy commented:
@kramasamy I think you're right that the 2 configs is overkill, I was mirroring the structure of the code (conf dir + file) but I think it makes more sense just to have one config. I've updated the code + added some documentation, I've also changed the conf files to call out that the line is commented out and what it will do more clearly.

billonahill commented:
+1 for not requiring javadocs on unit test if you're able to add that inclusion. Best to do so in a separate PR please.

kramasamy commented:
@chris-pardy - can you please fill in the CLA?  https://engineering.twitter.com/opensource/cla

chris-pardy commented:
@kramasamy I just filled out the CLA. Is there a way I can get this PR to  build again?

kramasamy commented:
@chris-pardy - no idea why the ci did not kick in. Try another commit and update the pull request?

kramasamy commented:
@chris-pardy - some Java style issue, "unused import"

kramasamy commented:
This seems to be simple fix - 

```
git diff
diff --git a/heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraScheduler.java b/heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraScheduler.java
index 64ecffb..82db52c 100644
--- a/heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraScheduler.java
+++ b/heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraScheduler.java
@@ -14,7 +14,6 @@
 
 package com.twitter.heron.scheduler.aurora;
 
-import java.io.File;
 import java.nio.charset.Charset;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -34,7 +33,6 @@ import com.twitter.heron.common.basics.FileUtils;
 import com.twitter.heron.proto.scheduler.Scheduler;
 import com.twitter.heron.scheduler.UpdateTopologyManager;
 import com.twitter.heron.spi.common.Config;
-import com.twitter.heron.spi.common.ConfigKeys;
 import com.twitter.heron.spi.common.Context;
 import com.twitter.heron.spi.common.Misc;
 import com.twitter.heron.spi.packing.PackingPlan;
```

But after fixing the checkstyle, it seems to be failing in the unit test

chris-pardy commented:
@kramasamy not sure what happened to the CI build but it doesn't look related to my PR?

kramasamy commented:
@chris-pardy - looks good. there were some flaky integration tests. we are taking a look at them.

kramasamy commented:
@chris-pardy - is this ready to go?

chris-pardy commented:
Yep

On Sun, Oct 9, 2016, 14:22 Karthik Ramasamy <notifications@github.com>
wrote:

> @chris-pardy <https://github.com/chris-pardy> - is this ready to go?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/twitter/heron/pull/1420#issuecomment-252503049>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFZplrnZKn6rptdKl25OTDJO6oSFWeN-ks5qyTDygaJpZM4KFYgP>
> .
>

kramasamy commented:
👍
